### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [2]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [3]:
data1.loc[data1['season'] == 2015].head()

id  season     city        date                  team1  \
517  518    2015  Kolkata  2015-04-08         Mumbai Indians   
518  519    2015  Chennai  2015-04-09    Chennai Super Kings   
519  520    2015     Pune  2015-04-10       Rajasthan Royals   
520  521    2015  Chennai  2015-04-11    Chennai Super Kings   
521  522    2015  Kolkata  2015-04-11  Kolkata Knight Riders   

                           team2                  toss_winner toss_decision  \
517        Kolkata Knight Riders        Kolkata Knight Riders         field   
518             Delhi Daredevils             Delhi Daredevils         field   
519              Kings XI Punjab              Kings XI Punjab         field   
520          Sunrisers Hyderabad          Chennai Super Kings           bat   
521  Royal Challengers Bangalore  Royal Challengers Bangalore         field   

     result  dl_applied                       winner  win_by_runs  \
517  normal           0        Kolkata Knight Riders            0   
518  normal           0          Chennai Super Kings            1   
519  normal           0             Rajasthan Royals           26   
520  normal           0          Chennai Super Kings           45   
521  normal           0  Royal Challengers Bangalore            0   

     win_by_wickets player_of_match                                    venue  \
517               7        M Morkel                             Eden Gardens   
518               0         A Nehra          MA Chidambaram Stadium, Chepauk   
519               0     JP Faulkner  Maharashtra Cricket Association Stadium   
520               0     BB McCullum          MA Chidambaram Stadium, Chepauk   
521               3        CH Gayle                             Eden Gardens   

            umpire1        umpire2 umpire3  
517          S Ravi  C Shamshuddin     NaN  
518  RK Illingworth    VA Kulkarni     NaN  
519          SD Fry    CB Gaffaney     NaN  
520  RK Illingworth    VA Kulkarni     NaN  
521          S Ravi  C Shamshuddin     NaN

### Replacing `Delhi Daredevils` with `Delhi Capitals`:

In [4]:
data1 = data1.replace("Delhi Daredevils", "Delhi Capitals")

### Filtering Delhi and KXIP matches only:

In [5]:
kxip_dc = data1.loc[((data1['team1'] == 'Kings XI Punjab') & (data1['team2'] == 'Delhi Capitals') | (data1['team1'] == 'Delhi Capitals') & (data1['team2'] == 'Kings XI Punjab'))]
kxip_dc.head(10)

id  season          city        date            team1            team2  \
14    15    2017         Delhi  2017-04-15   Delhi Capitals  Kings XI Punjab   
34    35    2017    Chandigarh  2017-04-30   Delhi Capitals  Kings XI Punjab   
72    73    2008    Chandigarh  2008-04-27   Delhi Capitals  Kings XI Punjab   
99   100    2008         Delhi  2008-05-17   Delhi Capitals  Kings XI Punjab   
119  120    2009     Cape Town  2009-04-19  Kings XI Punjab   Delhi Capitals   
160  161    2009  Bloemfontein  2009-05-15   Delhi Capitals  Kings XI Punjab   
176  177    2010    Chandigarh  2010-03-13  Kings XI Punjab   Delhi Capitals   
217  218    2010         Delhi  2010-04-11   Delhi Capitals  Kings XI Punjab   
258  259    2011         Delhi  2011-04-23   Delhi Capitals  Kings XI Punjab   
292  293    2011    Dharamsala  2011-05-15  Kings XI Punjab   Delhi Capitals   

         toss_winner toss_decision  result  dl_applied           winner  \
14    Delhi Capitals           bat  normal           0   Delhi Capitals   
34   Kings XI Punjab         field  normal           0  Kings XI Punjab   
72    Delhi Capitals           bat  normal           0  Kings XI Punjab   
99    Delhi Capitals           bat  normal           1  Kings XI Punjab   
119   Delhi Capitals         field  normal           1   Delhi Capitals   
160  Kings XI Punjab         field  normal           0  Kings XI Punjab   
176   Delhi Capitals         field  normal           0   Delhi Capitals   
217   Delhi Capitals           bat  normal           0  Kings XI Punjab   
258  Kings XI Punjab         field  normal           0   Delhi Capitals   
292   Delhi Capitals         field  normal           0  Kings XI Punjab   

     win_by_runs  win_by_wickets   player_of_match  \
14            51               0       CJ Anderson   
34             0              10    Sandeep Sharma   
72             0               4         SM Katich   
99             6               0  DPMD Jayawardene   
119            0              10        DL Vettori   
160            0               6             B Lee   
176            0               5         G Gambhir   
217            0               7         PP Chawla   
258           29               0         DA Warner   
292           29               0         PP Chawla   

                                                 venue          umpire1  \
14                                    Feroz Shah Kotla         YC Barde   
34   Punjab Cricket Association IS Bindra Stadium, ...         YC Barde   
72          Punjab Cricket Association Stadium, Mohali      RE Koertzen   
99                                    Feroz Shah Kotla   AV Jayaprakash   
119                                           Newlands        MR Benson   
160                                    OUTsurance Oval  HDPK Dharmasena   
176         Punjab Cricket Association Stadium, Mohali      BR Doctrove   
217                                   Feroz Shah Kotla        BF Bowden   
258                                   Feroz Shah Kotla         S Asnani   
292       Himachal Pradesh Cricket Association Stadium        Asad Rauf   

         umpire2 umpire3  
14   Nitin Menon     NaN  
34     CK Nandan     NaN  
72     I Shivram     NaN  
99   RE Koertzen     NaN  
119    SD Ranade     NaN  
160    IL Howell     NaN  
176       S Ravi     NaN  
217    AM Saheba     NaN  
258  RE Koertzen     NaN  
292   SL Shastri     NaN

## Question 1: Who will win the match?

#### Solution: Checking counts of winner:

In [6]:
kxip_dc['winner'].value_counts()/kxip_dc.shape[0] * 100

Kings XI Punjab    58.333333
Delhi Capitals     41.666667
Name: winner, dtype: float64

**Answer:** As per the above analysis, winning percentage of KXIP is **58.33** over DC.

## Question 2: What will KL Rahul's strike rate be in first 6 overs?

#### Solution: Observing runs of KL Rahul in each match:

### Loading "Deliveries" dataset:

In [7]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

### Replacing `Delhi Daredevils` with `Delhi Capitals`:

In [8]:
data2 = data2.replace("Delhi Daredevils", "Delhi Capitals")

### Filtering Data where batsman is KL Rahul and over is between 1 and 6:

In [9]:
rahul = data2.loc[(data2['batsman'] == 'KL Rahul') & ((data2['over'] >= 1) & data2['over'] <= 6)]
rahul_balls = pd.DataFrame(rahul.groupby('match_id')['ball'].count())
rahul_runs = pd.DataFrame(rahul.groupby('match_id')['batsman_runs'].sum())
rahul_data = pd.merge(rahul_balls, rahul_runs, on='match_id')
rahul_data['Strike Rate'] = 100 * rahul_data['batsman_runs']/rahul_data['ball']
rahul_data = rahul_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
rahul_data

Balls  Runs  Strike Rate
match_id                          
401          12    12   100.000000
429           4     8   200.000000
461          19    20   105.263158
466          28    27    96.428571
474          27    25    92.592593
477          40    46   115.000000
481           6     6   100.000000
487          18    18   100.000000
493          13    10    76.923077
500          15    14    93.333333
521           8     5    62.500000
525          32    44   137.500000
529           5     2    40.000000
531          21    24   114.285714
537           3     0     0.000000
541          28    25    89.285714
564          10    17   170.000000
572          26    25    96.153846
590          14    23   164.285714
592          10     7    70.000000
595          36    51   141.666667
603          29    51   175.862069
606          33    52   157.575758
611          35    38   108.571429
615          26    42   161.538462
617          54    68   125.925926
626           7    16   228.571429
632          23    38   165.217391
633           1     0     0.000000
636           9    11   122.222222
7895         16    51   318.750000
7901         30    48   160.000000
7905         22    37   168.181818
7909         21    18    85.714286
7911         28    61   217.857143
7915         15    23   153.333333
7918         26    32   123.076923
7927         20    24   120.000000
7931         56    86   153.571429
7933         70   101   144.285714
7937         31    68   219.354839
7941         15    22   146.666667
7943         64    99   154.687500
7949         12     8    66.666667
11140         4     4   100.000000
11142         5     1    20.000000
11145        58    72   124.137931
11149        11    15   136.363636
11309        47    55   117.021277
11313        53    71   133.962264
11315        66   104   157.575758
11319        15    18   120.000000
11323        47    53   112.765957
11328         9    12   133.333333
11333        29    44   151.724138
11339        60    84   140.000000
11343         7     2    28.571429
11346        36    75   208.333333

### Average Strike Rate of KL Rahul in 1 to 6 overs:

In [10]:
avg_rahul_sr = rahul_data['Strike Rate'].mean()
avg_rahul_sr

126.32135253319336

**Answer:** As per the above analysis, average Strike Rate of KL Rahul in first 6 overs is **124**.

## Question 3: What will be Shreya's Iyer's contribution to DC's total score?

#### Solution: Shreyas is playing from Delhi side since 2015; i.e. mtch_id 518. Observing match wise DC runs and Shreyas Iyer's runs:

In [11]:
df = data2.loc[(data2['match_id'] >= 518) & (data2['batting_team'] == 'Delhi Capitals')]
dc_runs = pd.DataFrame(df.groupby('match_id')['total_runs'].sum())
dc_runs = dc_runs.rename({'total_runs' : 'DC Runs'}, axis = 1)
shreyas_runs = data2.loc[(data2['match_id'] >= 518) & (data2['batsman'] == 'SS Iyer')]
shreyas_runs = pd.DataFrame(shreyas_runs.groupby('match_id')['batsman_runs'].sum())
shreyas_runs = shreyas_runs.rename({'batsman_runs' : "Sheyas' Runs"}, axis = 1)
dc_data = pd.merge(dc_runs, shreyas_runs, on = 'match_id')
dc_data["Shreyas' Contribution"] = dc_data["Sheyas' Runs"]/dc_data['DC Runs'] * 100
dc_data

DC Runs  Sheyas' Runs  Shreyas' Contribution
match_id                                              
519           149             7               4.697987
523           184            40              21.739130
528           169             6               3.550296
531           167            60              35.928144
535           146            31              21.232877
539           190            83              43.684211
543            95             0               0.000000
545           158            40              25.316456
548           119            54              45.378151
553           175             9               5.142857
556           152            19              12.500000
561           157             0               0.000000
565           120            70              58.333333
571           187            20              10.695187
578            98             0               0.000000
583           113             3               2.654867
587           192             0               0.000000
593           164            19              11.585366
602           186             0               0.000000
625           121             8               6.611570
7895          172            13               7.558140
7902          197            27              13.705584
7906          136             4               2.941176
7912          186            52              27.956989
7915          148            58              39.189189
7919          223            97              43.497758
7923          206            13               6.310680
7925          203            50              24.630542
7929          166            44              26.506024
7935          190             3               1.578947
7938          188            35              18.617021
7945          171            21              12.280702
7948          186             6               3.225806
11139         219            16               7.305936
11141         154            18              11.688312
11146         199            49              24.623116
11149         163            30              18.404908
11152         134            43              32.089552
11311         164            73              44.512195
11317         183             6               3.278689
11321         164            48              29.268293
11325         136             3               2.205882
11328         177            61              34.463277
11331         194             4               2.061856
11337         193            52              26.943005
11341          99            44              44.444444
11344         126            15              11.904762
11413         170             8               4.705882
11414         155            14               9.032258

### Average Shreyas' contribution:

In [12]:
dc_data["Shreyas' Contribution"].mean()

17.224109325410332

**Answer:** As per the above analysis, we see that Shreyas Iyer's contributon is around **17%** to total Delhi's score.

## Question 4: How many wickets will be taken by KXIP bowlers excluding that taken by Mohammed Shami, from the 31st ball to 120th ball?

#### Solution: Observing wicket  counts of KXIP bowlers excluding Mohammed Shami. Filtering the data between 6th and 20th Over.

In [18]:
kxip_wickets = data2.loc[(data2['bowling_team'] == 'Kings XI Punjab') & (data2['bowler'] != 'Mohammed Shami') & ((data2['over'] >= 6) & (data2['over'] <= 20))]
kxip_wickets = pd.DataFrame(kxip_wickets.groupby('match_id')['player_dismissed'].count())
kxip_wickets = kxip_wickets.rename({'player_dismissed' : 'No. of Wickets'}, axis = 1)
kxip_wickets

No. of Wickets
match_id                
4                      5
8                      1
11                     2
15                     6
19                     5
22                     2
26                     6
32                     3
35                     7
42                     7
46                     4
48                     5
50                     6
54                     1
61                     4
65                     1
69                     7
73                     6
77                     7
81                     2
84                     7
91                     2
94                     6
97                     5
100                    4
106                    9
108                    4
112                    7
116                    0
123                    1
127                    8
130                    3
135                    4
139                    7
142                    2
145                    4
149                    2
151                    4
157                    1
161                    7
164                    6
169                    8
177                    3
181                    2
186                    6
190                    7
193                    4
196                    3
201                    6
205                    3
208                    3
212                    1
215                    8
218                    7
225                    4
228                    2
239                    2
243                    2
249                    7
256                    5
259                    4
270                    0
273                    4
280                    5
284                    4
287                    7
290                    7
293                    8
296                    8
300                    2
311                    4
315                    4
321                    9
325                    5
330                    2
333                    3
336                    6
340                    6
343                    8
350                    4
354                    5
359                    6
366                    4
369                    3
371                    5
374                    4
387                    7
391                    0
398                    3
400                    7
405                    4
409                    4
412                    4
415                    2
421                    2
425                    3
429                    4
436                    1
440                    6
443                    5
448                    3
450                    7
460                    4
464                    4
466                    7
472                    8
475                    4
479                    3
486                    5
488                    7
491                    1
496                    5
502                    6
505                    3
509                    7
512                    9
514                    7
516                    5
517                    6
520                    5
524                    5
528                    4
532                    4
536                    6
542                    2
544                    5
548                    1
552                    3
557                    3
560                    8
564                    5
566                    4
569                    1
579                    4
583                    1
586                    6
589                    4
594                    5
597                    5
604                    7
608                    3
612                    5
615                    6
619                    7
622                    3
626                    3
629                    5
7895                   6
7901                   4
7905                   3
7909                   2
7911                   6
7915                   6
7918                   3
7927                   4
7931                   7
7933                   7
7937                   6
7941      

### Average wickets:

In [19]:
kxip_wickets['No. of Wickets'].mean()

4.417142857142857

## Question 5: What will be the status of Chris Gayle's batting durin KXIP innings?

#### Solution: Filtering data with CSK and Delhi Matches:

In [22]:
gayle = data2.loc[(data2['batsman'] == 'CH Gayle')]
gayle_balls = pd.DataFrame(gayle.groupby('match_id')['ball'].count())
gayle_runs = pd.DataFrame(gayle.groupby('match_id')['batsman_runs'].sum())
gayle_data = pd.merge(gayle_balls, gayle_runs, on='match_id')
gayle_data['Strike Rate'] = 100 * gayle_data['batsman_runs']/gayle_data['ball']
gayle_data = gayle_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
gayle_data

Balls  Runs  Strike Rate
match_id                          
1            23    32   139.130435
5             9     6    66.666667
12           28    22    78.571429
20           40    77   192.500000
27           18     7    38.888889
30           12     8    66.666667
42            2     0     0.000000
45            1     0     0.000000
55           39    48   123.076923
121          12    10    83.333333
123          29    44   151.724138
126          44    54   122.727273
132          12    12   100.000000
134          38    40   105.263158
138           7     7   100.000000
142          17    17   100.000000
191          65    75   115.384615
196          13    14   107.692308
200          22    30   136.363636
204           7     4    57.142857
208          44    88   200.000000
213          22    40   181.818182
217          17    34   200.000000
222           4     7   175.000000
227           3     0     0.000000
258          57   102   178.947368
263          14    26   185.714286
268          26    49   188.461538
280          49   107   218.367347
283          17    44   258.823529
288          44    70   159.090909
291          12    38   316.666667
296           7     0     0.000000
302          54    75   138.888889
304           9     8    88.888889
306          50    89   178.000000
307           3     0     0.000000
317           8     2    25.000000
320          40    68   170.000000
326           9     8    88.888889
329          49    81   165.306122
333          61    87   142.622951
338           8     4    50.000000
344          59    86   145.762712
350          44    71   161.363636
356          23    26   113.043478
360          59    82   138.983051
362          32    57   178.125000
367           8     6    75.000000
372          62   128   206.451613
376          10    27   270.000000
383          61    92   150.819672
388           6    11   183.333333
392          55    85   154.545455
396          11     4    36.363636
401          12    15   125.000000
407          49    49   100.000000
411          69   175   253.623188
417          20    18    90.000000
420          17    34   200.000000
426          21    21   100.000000
429          56    77   137.500000
432          15    13    86.666667
438           7     4    57.142857
441          37    33    89.189189
448          34    61   179.411765
452          14    28   200.000000
475           7    20   285.714286
481          20    27   135.000000
484          25    38   152.000000
488           7     4    57.142857
492          25    19    76.000000
495          23    22    95.652174
499          52    46    88.461538
503          20    14    70.000000
506           9     6    66.666667
522          57    96   168.421053
525          16    21   131.250000
534          31    10    32.258065
540          17    20   117.647059
543          42    62   147.619048
546           4    10   250.000000
550          12    21   175.000000
557          63   117   185.714286
562          13    13   100.000000
566          15    17   113.333333
568          11    35   318.181818
571           5     1    20.000000
574          28    27    96.428571
575          45    41    91.111111
580           5     1    20.000000
587           3     0     0.000000
606           7     7   100.000000
617           6     5    83.333333
620          13     6    46.153846
624          31    49   158.064516
626          35    73   208.571429
632           4     1    25.000000
633          14     9    64.285714
636          41    76   185.365854
7905         34    64   188.235294
7909         65   111   170.769231
7911         39    63   161.538462
7918         22    23   104.545455
7927         44    58   131.818182
7931         11     8    72.727273
7933          2     2   100.000000
7937         17    21   123.529412
7941         14    18   128.571429
7943         11    18   163.636364
7949          2     0     0.000000
11140        52    91   175.000000
11142        14    2

#### Getting mean of the Chris Gayle's strike rate:

In [23]:
gayle_data['Strike Rate'].mean()

126.91220166371855

**Answer:** Average strike rate of Chris Gayle is **127**.